### PR TITLE
fix(chat): fix switching back to chats is very slow

### LINKED ISSUE #5615: it takes a few seconds to switch from setting to chats
https://user-images.githubusercontent.com/11790366/166627346-4dee9b6e-3b50-4a0f-b2f5-90542a7864bb.mov

varsion: [StatusIm-Desktop-220503-160728-952f39.dmg](https://ci.status.im/job/status-desktop/job/branches/job/macos/job/master/lastSuccessfulBuild/artifact/pkg/StatusIm-Desktop-220503-160728-952f39.dmg)

install app from scratch , open a few chats

1 open settings
2 press chats button
3 wait a few seconds (~2-3s)
4 chats are opened
5 memory usage jumps

### PR DESCRIPTION
**Pull request by jrainville**
Fixes #5615

This fixes some QML warnings that I thought might have a effect, but also makes the message loading more "async".

The fix I did is a bit hackish, but what is does is only show the first 10 messages in the first 10 milliseconds, and then enable the rest.

The reason why it was so slow was that the scroll goes wild when switching. I haven't been able to find the reason for that sadly.
Anyway, since the scroll goes crazy, it loads messages that shouldn't be loaded, so that's why it's slow.

Now, by loading only 10 at the start and waiting a fraction of a second to "load" the rest, the scroll has time to settle and switching is faster.

I tested on Windows and it made a good difference. The switch is not instantaneous, but on Windows it was always a bit slow. I would think that on Linux and Mac it should be very fast.

### CODE REVIEW

**Review by elina2015 — APPROVED** [review]

File: ui/app/AppLayouts/Chat/views/ChatMessagesView.qml [hunk 1]
```diff
@@ -237,12 +252,15 @@ Item {
...
         model: messageStore.messagesModel
 
-        // Not Refactored Yet
-        //Component.onCompleted: scrollToBottom(true)
+        Component.onCompleted: chatLogView.scrollToBottom(true)
 
         delegate: MessageView {
             id: msgDelegate
 
+            // We wait to load all messages, because switching back to chats makes
```
Comment (stefandunca) [inline ui/app/AppLayouts/Chat/views/ChatMessagesView.qml:260]: Wondering why is this required. The model-view-delegate in QML takes care of efficiently instantiating and destroying delegates based on the item count in the model and without the need to hide them.

File: ui/app/AppLayouts/Chat/views/ChatMessagesView.qml [hunk 2]
```diff
@@ -237,12 +252,15 @@ Item {
...
+            // the scroll go crazy so it loads way too many messages making it slow
+            visible:  index < 10 || root.allMessagesLoaded
```
Comment (stefandunca) [inline ui/app/AppLayouts/Chat/views/ChatMessagesView.qml:262]: Not sure how limiting to 10 visible delegates helps. One issue that I see and might be the root cause is that we don't define a delegate height and ListView has no idea how many delegates to instantiate. Do we somehow define the `implicitHeight` or `heigh` for a delegate somewhere where I didn't find it?

File: ui/app/AppLayouts/Chat/views/ChatMessagesView.qml [hunk 3]
```diff
@@ -42,6 +43,20 @@ Item {
...
+    property bool allMessagesLoaded: false
+
+    onIsActiveChannelChanged: {
+        if (!isActiveChannel) {
+            root.allMessagesLoaded = false
+            return
+        }
+        // We wait to load all messages, because switching back to chats makes
+        // the scroll go crazy so it loads way too many messages making it slow
+        timer.setTimeout(function() {
```
Comment (stefandunca) [inline ui/app/AppLayouts/Chat/views/ChatMessagesView.qml:55]: Maybe I miss things, but this looks like a workaround. The data model should handle the async loading and the asynchronous process be transparent to the View. @jrainville, how about a follow-up task to refactor this, or do you think it is final?

**Review by stefandunca — APPROVED** [review]
Good job for the quick fix. There is room for improvements regarding the async handling of loading messages and how we handle lazy loading. I hope we can do it soon.

File: ui/app/AppLayouts/Chat/views/ChatMessagesView.qml (hunk 3 above)
Comment (jrainville) [reply ui/app/AppLayouts/Chat/views/ChatMessagesView.qml:55]: > The data model should handle the async loading and the asynchronous process be transparent to the View

Yeah, I know, but somehow, the ListView was loading way too many items because, for reasons I don't know, it was moving the scroll when switching. This made it so that lots of rendering was done for no reason.

> how about a follow-up task to refactor this, or do you think it is final?

This shouldn't be final, but I couldn't find another solution as of now. Maybe when we do refactor the QML, we'll be rid of the problem or find a cleaner solution.

File: ui/app/AppLayouts/Chat/views/ChatMessagesView.qml (hunk 2 above)
Comment (endulab) [inline ui/app/AppLayouts/Chat/views/ChatMessagesView.qml:262]: @jrainville there are 2 properties which may help with delegates management: ListView::reuseItems and ListView::cacheBuffer. Did you play with these ? They may help and seems to be more elegant solution.

**Review by endulab — CHANGES_REQUESTED** [review]
